lldb/test/Shell/ObjectFile/ELF/elf-memory.test

```diff
@@ -0,0 +1,58 @@
+// REQUIRES: system-linux, native
```

Comment:
I would like to see a test which uses an on-disk section-header-free object file. (Maybe you could base it off of llvm/test/tools/llvm-readobj/ELF/dynamic-reloc-no-section-headers.test). The in-memory test has a lot of moving parts, it seems pretty fragile(*), and can only run on linux (or at least an elf system). OTOH, an on-disk test would be very self contained and can run anywhere. If this breaks (which I'm sure it eventually will), I'd much rather be debugging the second one.

I know that in-memory object files are your main use case, and we should have a test for that, but I think it should be like a cherry on top rather than the first&only line of defense. 

(*) It relies on certain entries being produces by the linker. Most of these are pretty much guaranteed to be there, but for example DT_NEEDED(libc) may not exist if the c library is linked statically. If we can test all of the small pieces with files, then I think it'd be best if the overall in memory test was a full end-to-end API test, which actually deletes the object file, and then verifies that we're still able to read certain things from the object.

Reply:
Adding support for memory ELF files is my primary concern with this patch as you know. Testing that in memory ELF files work as well as they can and that we can find the DT_DEBUG is the primary goal of this patch, so I would still like to concentrate on that. I have yet to run into an ELF file in the wild that doesn't have section headers, though I know they are possible. Testing I can load a ELF file from disk with no section headers is a whole other code path that introduces changes and time to this patch which I would rather not take on at this point if possible. Getting core files to be able to have a shared library list without having the main executable makes ELF core file actually useful now instead of being useless unless you get the correct executable. So although testing an on disk ELF file with no section headers is nice for testing, it can't replace the test I have written because we need this to test the in memory ELF files and testing the file itself is whole other code path that isn't needed.

I will be happy to add support for on disk ELF files after this patch, but I would like to get the useful part of this patch checked in so we can get ELF core files to be useful with no executable.

Reply:
If it was a lot of work then sure, I'd agree with you -- but I don't think that's the case here. I just gave it a shot, and this is the only change I needed to make on top of your patch to be able to parse DT_NEEDED of a section-free file (disclaimer: This code is mainly for demonstration, it's possible there is a better way to do that):
```
diff --git a/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp b/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp
index d26215a9ce3f..5f0d2233b865 100644
--- a/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp
+++ b/lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp
@@ -3810,28 +3810,26 @@ std::optional<DataExtractor> ObjectFileELF::GetDynstrData() {
     }
   }
 
-  // Every ELF file which represents an executable or shared library has
-  // mandatory .dynamic entries. Two of these values are DT_STRTAB and DT_STRSZ
-  // and represent the dynamic symbol tables's string table. These are needed
-  // by the dynamic loader and we can read them from a process' address space.
-  //
-  // When loading and ELF file from memory, only the program headers end up
-  // being mapped into memory, and we can find these values in the PT_DYNAMIC
-  // segment.
-  if (!IsInMemory())
-    return std::nullopt;
-  ProcessSP process_sp(m_process_wp.lock());
-  if (!process_sp)
-    return std::nullopt;
-
   const ELFDynamic *strtab = FindDynamicSymbol(DT_STRTAB);
   const ELFDynamic *strsz = FindDynamicSymbol(DT_STRSZ);
   if (strtab == nullptr || strsz == nullptr)
     return std::nullopt;
 
-  if (DataBufferSP data_sp =
-          ReadMemory(process_sp, strtab->d_ptr, strsz->d_val))
-    return DataExtractor(data_sp, GetByteOrder(), GetAddressByteSize());
+  if (ProcessSP process_sp = m_process_wp.lock()) {
+    if (DataBufferSP data_sp =
+            ReadMemory(process_sp, strtab->d_ptr, strsz->d_val))
+      return DataExtractor(data_sp, GetByteOrder(), GetAddressByteSize());
+  } else {
+    Address addr;
+    if (addr.ResolveAddressUsingFileSections(strtab->d_ptr, GetSectionList())) {
+      DataExtractor data;
+      addr.GetSection()->GetSectionData(data);
+      return DataExtractor(data,
+                           strtab->d_ptr - addr.GetSection()->GetFileOffset(),
+                           strsz->d_val);
+    }
+  }
+
   return std::nullopt;
 }
 
```

And I think this goes a long way towards improving the testability of this patch. I know section-free files aren't likely to be of much use, but they're really great for testing, and in terms of raw LOC, such a test may cover 80% of this patch.

Reply:
I added a test with an on disk file